# Predictions

### Data Analysis

In [4]:
import seaborn as sns
import numpy as np
import pandas as pd

weather=pd.read_csv('weatherAUS.csv')


In [5]:
import matplotlib.pyplot as plt
%matplotlib inline

In [6]:
weather.keys()

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow'],
      dtype='object')

In [7]:
weather['RainTomorrow'][55:100]
#weather.RainTomorrow.replace(to_replace=['No', 'Yes'], value=[0, 1],inplace=True)
#weather.RainToday.replace(to_replace=['No', 'Yes'], value=[0, 1],inplace=True)

55     No
56     No
57     No
58     No
59     No
60     No
61     No
62     No
63     No
64     No
65     No
66     No
67     No
68     No
69     No
70     No
71     No
72     No
73    Yes
74     No
75     No
76     No
77     No
78     No
79     No
80     No
81     No
82     No
83     No
84     No
85     No
86     No
87     No
88     No
89     No
90     No
91     No
92     No
93     No
94     No
95     No
96     No
97     No
98     No
99    Yes
Name: RainTomorrow, dtype: object

In [8]:
dummy=pd.get_dummies(weather['RainTomorrow'])
weather['RainTomorrow']

0          No
1          No
2          No
3          No
4          No
         ... 
145455     No
145456     No
145457     No
145458     No
145459    NaN
Name: RainTomorrow, Length: 145460, dtype: object

In [9]:
weather['RainTomorrow'][0]

'No'

In [10]:
weather.head()


Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

### Fitting

In [11]:
from sklearn.model_selection import train_test_split
weather.drop(labels=['Date','Location','WindDir9am','WindGustDir','WindDir3pm'],axis=1,inplace=True)
#weather.drop(labels=['WindDir3pm'],axis=1,inplace=True)
weather.RainTomorrow.replace(to_replace=['No', 'Yes'], value=[0, 1],inplace=True)
weather.RainToday.replace(to_replace=['No', 'Yes'], value=[0, 1],inplace=True)

weather=weather.fillna(0)

X=weather.drop('RainTomorrow',axis=1)
y=weather['RainTomorrow']
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.30)


In [ ]:
weather


# Logistic Regression

In [12]:
from sklearn.linear_model import LogisticRegression
logmodel = LogisticRegression()
logmodel.fit(X_train,y_train)
predictions = logmodel.predict(X_test)
from sklearn.metrics import confusion_matrix,classification_report
print(confusion_matrix(y_test,predictions))
print(classification_report(y_test,predictions))

[[32311  1853]
 [ 5811  3663]]
              precision    recall  f1-score   support

         0.0       0.85      0.95      0.89     34164
         1.0       0.66      0.39      0.49      9474

    accuracy                           0.82     43638
   macro avg       0.76      0.67      0.69     43638
weighted avg       0.81      0.82      0.81     43638



/Users/aman/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## **K-nearest Neighbours**

In [13]:
from sklearn.neighbors import KNeighborsClassifier


*choosing k value*

In [ ]:
error_rate = []

for i in range(1,40):
    
    knn = KNeighborsClassifier(n_neighbors=i)
    knn.fit(X_train,y_train)
    pred_i = knn.predict(X_test)
    error_rate.append(np.mean(pred_i != y_test))
    


In [ ]:
error_rate

In [ ]:
error_rate.index(min(error_rate))

In [ ]:
knn = KNeighborsClassifier(n_neighbors=34)

knn.fit(X_train,y_train)
pred = knn.predict(X_test)

print('WITH K=34')
print('\n')
print(confusion_matrix(y_test,pred))
print('\n')
print(classification_report(y_test,pred))

## Desicion tree

In [ ]:
from sklearn.tree import DecisionTreeClassifier
dtree = DecisionTreeClassifier()
dtree.fit(X_train,y_train)
predictions = dtree.predict(X_test)
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_test,predictions))
print(classification_report(y_test,predictions))

## **Random Forests**

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_estimators=100)
rfc.fit(X_train, y_train)


In [ ]:
rfc_pred = rfc.predict(X_test)
print(confusion_matrix(y_test,rfc_pred))


In [ ]:
print(classification_report(y_test,rfc_pred))

## **SVM**


In [10]:
from sklearn.svm import SVC

In [ ]:
svc_model=SVC()
svc_model.fit(X_train,y_train)

In [ ]:
from sklearn.model_selection import GridSearchCV

In [13]:
param_grid = {'C': [1, 10], 'gamma': [0.1,0.01]} 

In [14]:
grid=GridSearchCV(SVC(),param_grid,verbose=2,refit=2)

In [ ]:
grid.fit(X_train,y_train)

Fitting 5 folds for each of 4 candidates, totalling 20 fits
[CV] C=1, gamma=0.1 ..................................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ................................... C=1, gamma=0.1, total=26.9min
[CV] C=1, gamma=0.1 ..................................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed: 26.9min remaining:    0.0s


[CV] ................................... C=1, gamma=0.1, total=26.8min
[CV] C=1, gamma=0.1 ..................................................
[CV] ................................... C=1, gamma=0.1, total=26.9min
[CV] C=1, gamma=0.1 ..................................................


In [ ]:
grid_predictions=grid.predict(X_test)

In [ ]:
print(confusion_matrix(y_test,grid_predictions))

In [ ]:
print(classification_report(y_test,grid_predictions))

## **K means Clustering**

In [ ]:
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=2)
kmeans.fit(weather.drop('RainTomorrow',axis=1))

In [ ]:
#kmeans.cluster_centers_

In [ ]:
def converter(cluster):
    if cluster==0:
        return 0
    else:
        return 1

In [ ]:
weather['Cluster'] = weather['RainTomorrow'].apply(converter)

In [ ]:
from sklearn.metrics import confusion_matrix,classification_report
print(confusion_matrix(weather['Cluster'],kmeans.labels_))
print(classification_report(weather['Cluster'],kmeans.labels_))

# Results:

## **Logistic regression**
[[32311  1853]
 [ 5811  3663]]


## **KNN:**

K=34

[[32726  1447]

 [ 5165  4300]]


## **Random Forests:**
trees=100

[[32632  1541]

 [ 4794  4671]]

## **SVM:**



## **KMeansClustering:** 

[[101460  12123]

 [ 28718   3159]]
